src/middlewared/middlewared/plugins/smb.py

```diff
@@ -72,6 +72,125 @@ def key_cp(encoding):
+                sid_is_present = True
+            if group.rsplit('-', 1)[-1] != "512" and group != sid:
+                self.logger.debug(f"Removing {group} from local admins group.")
+                rem = await Popen(
+                    ['/usr/local/bin/net', 'groupmap', 'delmem', 'S-1-5-32-544', group],
+                    stdout=subprocess.PIPE, stderr=subprocess.PIPE
+                )
+                remout = await rem.communicate()
+                if rem.returncode != 0:
+                    raise CallError(f'Failed to remove sid [{sid}] from S-1-5-32-544: {remout[1].decode()}')
+
+        if sid_is_present:
+            return False
+        else:
+            return True
+
+    @private
+    async def wbinfo_gidtosid(self, gid):
+        verrors = ValidationErrors()
+        proc = await Popen(
+            ['/usr/local/bin/wbinfo', '--gid-to-sid', f"{gid}"],
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        output = await proc.communicate()
+        if proc.returncode != 0:
+            if "WBC_ERR_WINBIND_NOT_AVAILABLE" in output[1].decode():
+                return "WBC_ERR_WINBIND_NOT_AVAILABLE"
+            else:
+                verrors.add('smb_update.admin_group', f"Failed to identify Windows SID for group: {output[1].decode()}")
+                raise verrors
+
+        return output[0].decode().strip()
+
+    @private
+    async def add_admin_group(self, admin_group=None, check_deferred=False):
+        """
+        Add a local or directory service group to BUILTIN\\Administrators (S-1-5-32-544)
+        Members of this group have elevated privileges to the samba server (ability to
+        take ownership of Files, override ACLs, view and modify user quotas, and administer
+        the server via the Computer Management MMC Snap-In. Unfortuntely, group membership
+        must be managed via "net groupmap listmem|addmem|delmem", which requires that
+        winbind be running when the commands are executed. In this situation net command
+        will fail with WBC_ERR_WINBIND_NOT_AVAILABLE. If this error message is returned, then
+        flag for a deferred command retry when service starts.
+
+        @param-in (admin_group): This is the group to add to BUILTIN\\Administrators. If unset, then
+            look up the value in the config db.
+        @param-in (check_deferred): If this is True, then only perform the group mapping if this has
+            been flagged as in need of deferred setup (i.e. samba wasn't running when it was initially
+            called). This is to avoid unecessarily calling during service start.
+        """
+
+        verrors = ValidationErrors()
+        if check_deferred:
+            is_deferred = await self.middleware.call('cache.has_key', 'SMB_SET_ADMIN')
+            if not is_deferred:
+                self.logger.debug("No delayed action to add admin_group detected.")
```

Comment:
Not clear what "detected" means here. Does this sentence mean "Did not find delayed action to add admin group." or "There is no delayed action to add a detected admin_group."?